SQL: relax version lock between server and clients

bpintea opened the pull request:
This PR is to allow older-than-server clients to connect, if these are past or on a
certain min release, `7.7.0`.
Note: the check is currently lower bounded only and subject to later adjustment to `major minus one` limitation.

bpintea committed:
Relax version lock between ES/SQL and clients

Allow older-than-server clients to connect, if these are past or on a
certain min release.

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

costin reviewed (approved):
LGTM

matriv commented on x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/VersionParityTests.java:
```diff
@@ -38,14 +38,19 @@ public void testExceptionThrownOnIncompatibleVersions() throws IOException, SQLE
+        do {
+            prepareResponse(version);
+
+            String url = JdbcConfiguration.URL_PREFIX + webServerAddress();
```
Those can be outside of the loop no?
and probably the whole try - catch can "wrap" the loop?

bpintea replied:
Yes, you're right, thanks!

matriv reviewed (approved):
LGTM. Left a minor comment.

astefan commented on x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/AbstractSqlQueryRequest.java:
```diff
@@ -233,6 +234,11 @@ public ActionRequestValidationException validate() {
         return validationException;
     }
 
+    protected boolean isClientCompatible() {
```
Do you think this method better belongs to `SqlVersion` class where `hasVersionCompatibility()` is also defined?

bpintea replied:
Yeh, I had a hesitation initially, but it does belong there more naturally. Thanks for raising it.

astefan reviewed (approved):
LGTM. Left one question, though.

bpintea committed:
Adress review comments

- refactor VersionParityTests#testNoExceptionThrownForCompatibleVersions
to take idempotent assignments and try-catch blocks out a loop.
- move method that establishes version compatibiltiy into SqlVersion.

matriv reviewed (approved):
LGTM

bpintea commented:
@elasticmachine run elasticsearch-ci/bwc

bpintea commented:
@elasticmachine run elasticsearch-ci/default-distro

bpintea commented:
@elasticmachine update branch

elasticmachine commented:
user doesn't have permission to update head repository

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/lift_version_lock_step
